### ISSUE TITLE
UnusedLocalVariable FP for var declared in a try-with-resource block

### ISSUE THREAD

**Issue by garydgregory** [pmd/pmd#4987]
Affects PMD Version: 7.1.0

Rule: UnusedLocalVariable

Please provide the rule name and a link to the rule documentation:
UnusedLocalVariable
https://pmd.github.io/pmd/pmd_rules_java_bestpractices.html#unusedlocalvariable

Description:

You MUST declare a local variable in a try-with-resources block, whether the block uses it or not. This happens in cases where a resource must be managed but need not be referred.

For example,
```java
        try (var myObject = MyFactory.createManagedResource(a, b, c)) {
            ...
            // Do stuff without referring to `myObject`.
            ...
        }
        // myObject is now properly closed.
```

**Expected outcome:**

PMD reports a violation at line for `myObject`, but that's wrong. That's a false positive.

Running PMD through: Maven

**Comment by jsotuyod** [comment]
This is actually by design. The var is unused, and a judgment call needs to be made wether the resource needs to be kept (ie: side-effects) or not. We can't simply assume it is needed blindly.

If you want to keep it, even if unused, you should rename the variable appropriately to use either `unused`,  `ignored` or even better under newer Java versions, `_`

**Comment by garydgregory** [comment]
@jsotuyod I feel you're missing the fact that you MUST declare the variable, the compiler will give an error otherwise. The resource MUST be managed and the try-with-resources block does just that. 

I would say that the compiler in fact _does use_ the variable in order to call the `close()` method. 

What PMD complains about then is that the variable is used by the compiler but not _visisble_ in the source code. Hence my view this is a false positive.

**Comment by oowekyala** [comment]
The variable is not used by any explicit code, so it is unused in that sense. If you see an unused variable you have to do a double take to find out whether that's expected or not. The point of renaming it is to make it clear to someone reading the code that this variable is _intentionally_ unused.

**Comment by garydgregory** [comment]
There is no double-take when you know what the Java syntax requires; _especially_ since the compiler does not give any other option.

PMD is making an incorrect assumption in an attempt at hand-holdibg the user IMO, which is painful since there is no reasonable remedy for a user to act upon.

**Comment by oowekyala** [comment]
The Java compiler also gives no other option but to name exception parameters in a catch block. Yet nobody complains that reporting an unused exception variable is a false positive. The point is a reader of the code might wonder if the variable is intentionally unused or not. If you make that clear through the variable name that is not the case. Renaming an unused variable is trivial and is a reasonable remedy.

**Comment by jsotuyod** [comment]
Also, that's the very reason why Java 21 introduced unnamed variables as a preview feature. The use is completely legit, but if it's not expected to be used, make it explicit.

Let's assume there is a try-with-resources block somewhere, and after a refactor the resource is not used anymore (although the exception handlers may still be applicable). Is the leftover resource needed or an omission? We can't tell, but there is definitely value in PMD letting them know if it's an omission. PMD needs to make a call. Ignoring all resources means erring on the side of false-negatives. Flagging them, mean erring on the side of false-positives. Given the explicit naming escape hatch, we feel we provide a reasonable compromise.

**Comment by garydgregory** [comment]
I think the "_" name workaround is the best idea, but 50% of users according to various surveys are still on Java 8, so that does not help and unfortunately most of the Aapche Commons components I maintain are on Java 8.

I'll have to decide whether to use PMD exclusion files or misleading  names like "unused" for those components.

TY for bouncing ideas around!
